core/src/builder.rs

```diff
@@ -92,5 +92,5 @@ pub trait TxBuilder: std::marker::Sized {
     fn locktime(self, locktime: u32) -> Self;
 
     /// Consume the builder and produce a transaction from the builder's current state.
-    fn build(self) -> Self::Transaction;
+    fn build(self) -> Result<Self::Transaction, <Self::Transaction as Transaction>::TxError>;
```

Comment:
it'd be nice if we could make this return the builder in the error case, so it could be retried. thoughts?

Reply:
Something like `Result<Self::Transaction, (Self, TxError)>`

it's not pretty tho :thinking:

Reply:
The consumer of the builder would need to be really sure of the state of the builder after getting it back. The `TxError` will tell them what is wrong, but you don't want to accidentally add the same input twice or an extra output (potentially huge tx fee).

Another option that I don't love is adding a function to the builder that you can check to guarantee there will not be a `TxError`.

```rust
fn is_buildable(self) -> bool
```

Or leaving that function and then adding a new one:

```rust
fn try_build(self) -> Result<Self::Transaction, (Self, TxError)>
```

That makes the API larger so idk if I like that one either.

Also need to take into consideration the wasm binding, the error case is a `JsValue` and ends up being a thrown error on the JS side. Would need `JsValue::from((Self, TxError))` or have the API be slightly different

```js
let builder = new Builder()

try {
  let tx = builder.build()
} catch (e) {
  console.log('error here')
}
```

Willing to add a commit that updates to returning `Result<Self::Transaction, (Self, TxError)>`

Reply:
I think on reflection it doesn't really work 🤔

the caller can always clone the builder if they want to

Reply:
I like the `is_buildable` is an interesting idea. Could we offload that check to the tx and not make any allocations tho? Lemme think more